blog/2019-03-28-debezium-at-pipedrive.adoc

```diff
@@ -0,0 +1,225 @@
+
+Example connector job template:
+
+[source,json]
+----
+{
+ "tasks.max": "1",
+ "connector.class": "io.Debezium.connector.MySQL.MySQLConnector",
+
+ "database.hostname": "{MySQL.host}",
+ "database.port": "{MySQL.port}",
+ "database.user": "{MySQL.username}",
+ "database.password": "{MySQL.password}",
+ "database.server.id": "{Debezium.server.id}",
+ "database.server.name": "company_db_{baseID}",
+
+ "database.history.kafka.bootstrap.servers": "{kafka.servers}",
+ "database.history.kafka.topic": "Debezium-history-company-db{baseID}",
+ "database.history.store.only.monitored.tables.ddl": true,
+ "database.history.skip.unparseable.ddl": "true",
+
+ "ddl.parser.mode": "antlr",
+ "snapshot.mode": "schema_only",
+
+ "snapshot.locking.mode": "none",
+ "only.in.env.test.snapshot.locking.mode": "minimal",
+
+ "gtid.new.channel.position": "earliest"
+}
+----
+
+== Connector health monitoring and automatic restart.
+
+As Debezium connector can’t automatically recover from MySQL connection issues, it needs another management script to restart the task. At first we used external cron like script for this, but now as nodejs is already running in container, we delegated this responsibility there as well.
+
+The logic is very simple - every 30 seconds it connects to the Kafka Connect REST API running in the local container and retrieves the connector status. If the connector or task is in FAILED state, it tries to restart it by posting connectors/jobname/tasks/1/restart up to 5 times. If still doesn’t work or there’s UNASSIGNED state, starts graceful shutdown of the Docker container.
+
+== Securing the Kafka Connect REST API
+Because the Kafka Connect REST API had no built in security (in the beginning of 2018 at least), anybody could query and see Kafka cluster and MySQL credentials in plaintext. To prevent such leak we’ve added simple nodejs http proxy script with http basic auth to different port and only expose this port to outside. Once auth is passed this nodejs proxy just forwards all the requests to kafka connect local port 8080.
+
+== Monitoring with Prometheus/Grafana
+Debezium exposes nice JMX variables, but to query them using prometheus scraper we added prometheus exporter into Debezium docker image. This library has worked without problems https://github.com/prometheus/jmx_exporter[jmx_prometheus_javaagent] and we’re most Debezium and kafka connect prefixed variables as prometheus metrics.
+
+Our main alerts are based on https://debezium.io/docs/connectors/mysql/#binlog-metrics[MySQL_binlog_connected metric], but unfortunately it hasn’t proven absolutely reliable, so there is additional alerts based on logs, counting “org.apache.kafka.connect.errors.ConnectException” exceptions.
```

Comment:
Do you know what the problem there is? Can you file an issue in DBZ for investigating this?